StringIndexOutOfBoundsException when using TO_CHAR

Hi all,

Using H2 1.4.200 (2019-10-14), if you execute the following:
`SELECT TO_CHAR(CAST(10000000 AS DOUBLE), 'fm999,999,999.99')`

... then you get the exception:
```
Exception in thread "main" org.h2.jdbc.JdbcSQLNonTransientException: General error: "java.lang.StringIndexOutOfBoundsException: String index out of range: 7" [50000-200]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:505)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:429)
	at org.h2.message.DbException.get(DbException.java:194)
	at org.h2.message.DbException.convert(DbException.java:347)
	at org.h2.message.DbException.toSQLException(DbException.java:319)
	at org.h2.message.TraceObject.logAndConvert(TraceObject.java:366)
	at org.h2.jdbc.JdbcStatement.executeQuery(JdbcStatement.java:100)
	at test.h2sioob.H2StringIndexOutOfBoundsTest.main(H2StringIndexOutOfBoundsTest.java:18)
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: 7
	at java.base/java.lang.StringLatin1.charAt(StringLatin1.java:47)
	at java.base/java.lang.String.charAt(String.java:702)
	at org.h2.expression.function.ToChar.toChar(ToChar.java:264)
	at org.h2.expression.function.Function.getValueWithArgs(Function.java:1468)
	at org.h2.expression.function.Function.getValue(Function.java:672)
	at org.h2.expression.function.Function.optimize(Function.java:2888)
	at org.h2.command.dml.Select.prepare(Select.java:1206)
	at org.h2.command.Parser.prepareCommand(Parser.java:744)
	at org.h2.engine.Session.prepareLocal(Session.java:657)
	at org.h2.engine.Session.prepareCommand(Session.java:595)
	at org.h2.jdbc.JdbcConnection.prepareCommand(JdbcConnection.java:1235)
	at org.h2.jdbc.JdbcStatement.executeQuery(JdbcStatement.java:78)
	... 1 more
```

If you change the number with a smaller number, then it works.
If you change the cast to NUMBER, then it works.

Here is a test case:
[H2StringIndexOutOfBoundsTest.zip](https://github.com/h2database/h2database/files/4192816/H2StringIndexOutOfBoundsTest.zip)

This looks similar to issue #115

Are we doing something wrong?

--- a/h2/src/docsrc/html/changelog.html
+++ b/h2/src/docsrc/html/changelog.html
@@ -21,6 +21,8 @@ Change Log
 
 <h2>Next Version (unreleased)</h2>
 <ul>
+<li>Issue #2439: StringIndexOutOfBoundsException when using TO_CHAR
+</li>
 <li>Issue #2444: WHEN NOT MATCHED THEN INSERT should accept only one row
 </li>
 <li>Issue #2434: Next value expression should return the same value within a processed row

--- a/h2/src/main/org/h2/expression/function/ToChar.java
+++ b/h2/src/main/org/h2/expression/function/ToChar.java
@@ -237,8 +237,11 @@ public class ToChar {
         // go ahead and do that first
         int separator = findDecimalSeparator(format);
         int formatScale = calculateScale(format, separator);
-        if (formatScale < number.scale()) {
+        int numberScale = number.scale();
+        if (formatScale < numberScale) {
             number = number.setScale(formatScale, RoundingMode.HALF_UP);
+        } else if (numberScale < 0) {
+            number = number.setScale(0);
         }
 
         // any 9s to the left of the decimal separator but to the right of a

--- a/h2/src/test/org/h2/test/db/TestFunctions.java
+++ b/h2/src/test/org/h2/test/db/TestFunctions.java
@@ -1837,6 +1837,9 @@ public class TestFunctions extends TestDb implements AggregateFunction {
         assertResult(twoDecimals, stat, "select to_char(0, 'FM0D009') from dual;");
         assertResult(oneDecimal, stat, "select to_char(0, 'FM0D09') from dual;");
         assertResult(oneDecimal, stat, "select to_char(0, 'FM0D0') from dual;");
+
+        assertResult("10,000,000.", stat,
+                "SELECT TO_CHAR(CAST(10000000 AS DOUBLE PRECISION), 'FM999,999,999.99') FROM DUAL");
         conn.close();
     }
 

